edgelet/iotedged/src/lib.rs

```diff
@@ -571,7 +582,104 @@ where
     }
 }
 
-fn set_iot_edge_env_vars<S>(settings: &S) -> Result<(), Error>
+fn retrieve_external_provisioning_info(
+    external: &External,
+    tokio_runtime: &mut tokio::runtime::Runtime,
+) -> Result<ProvisioningResult, Error> {
+    info!("Retrieving provisioning information from the external endpoint...");
+    let external_provisioning_client = ExternalProvisioningClient::new(external.endpoint())
+        .context(ErrorKind::Initialize(
+            InitializeErrorReason::ExternalProvisioningClient(
+                ExternalProvisioningErrorReason::ClientInitialization,
+            ),
+        ))?;
+    let external_provisioning = ExternalProvisioning::new(external_provisioning_client);
+
+    let provision_fut = external_provisioning
+        .provision(MemoryKeyStore::new())
+        .map_err(|err| {
+            Error::from(err.context(ErrorKind::Initialize(
+                InitializeErrorReason::ExternalProvisioningClient(
+                    ExternalProvisioningErrorReason::Provisioning,
+                ),
+            )))
+        });
+
+    tokio_runtime.block_on(provision_fut)
+}
+
+fn get_external_provisioning_info<S>(
+    settings: &S,
+    mut tokio_runtime: &mut tokio::runtime::Runtime,
+) -> Result<Option<ProvisioningResult>, Error>
+where
+    S: RuntimeSettings,
+{
+    if let Provisioning::External(ref external) = settings.provisioning() {
+        // Set the external provisioning endpoint environment variable for use by the custom HSM library.
+        env::set_var(
+            EXTERNAL_PROVISIONING_ENDPOINT_KEY,
+            external.endpoint().as_str(),
+        );
+
+        let prov_info = retrieve_external_provisioning_info(external, &mut tokio_runtime)?;
```

Comment:
`tokio_runtime` is already a `&mut`. You don't need to pass `&mut tokio_runtime`.

That will also allow you to change the parameter from `mut tokio_runtime: &mut` to `tokio_runtime: &mut`

Reply:
Doing that gives me the following error:

`error[E0308]: mismatched types                                                                                                                                                                                                                                                     --> iotedged\src\lib.rs:265:55                                                                                                                                                                                                                                                   |                                                                                                                                                                                                                                                                           265 |             get_external_provisioning_info(&settings, tokio_runtime)?;                                                                                                                                                                                                        |                                                       ^^^^^^^^^^^^^                                                                                                                                                                                                           |                                                       |                                                                                                                                                                                                                       |                                                       expected mutable reference, found struct 'tokio::runtime::Runtime'                                                                                                                                                      |                                                       help: consider mutably borrowing here: '&mut tokio_runtime'                                                                                                                                                            |                                                                                                                                                                                                                                                                               = note: expected type '&mut tokio::runtime::Runtime'                                                                                                                                                                                                                                      found type 'tokio::runtime::Runtime'`

Reply:
I'm telling you to change line 613 to

```
tokio_runtime: &mut tokio::runtime::Runtime,
```

and line 625 to

```
let prov_info = retrieve_external_provisioning_info(external, tokio_runtime)?;
```

Reply:
Done